### PR TITLE
added option to only sign the referenced element when passed reference_uri

### PR DESCRIPTION
**Pull request by soby**

### CODE REVIEW

**Comment by danimaribeiro** [comment]
Found one little problem:
When we have a namespace in the doc like:


`<tag xmlns="http://domain.com/namespace"></tag>`

It does not find the element.

I ended up doing something like this to work with empty prefix namespaces.

```
ns = doc_root.nsmap
if None in ns:
    ns['ns'] = ns[None]
    ns.pop(None)

targets = doc_root.xpath('//ns:*[@Id="{0}"]'.format(reference_uri), namespaces=ns)
```

Also there is the case where you have a prefix in the namespace:

`<tag xmlns:prefix="http://domain.com/namespace"></tag>`

Then I think something like this should resolve (didn't test this one). 

`targets = doc_root.xpath('//{0}:*[@Id="{1}"]'.format(doc_root.nsmap.keys()[0], reference_uri), namespaces=doc_root.nsmap)`

**Comment by soby-taulia** [comment]
Strange, I don't think I'm seeing that:

```
from lxml import etree as ET
x='<tag xmlns="http://domain.com/namespace" Id="blah"><other:something xmlns:other="http://other.com/other" /></tag>'
ET.fromstring(x).xpath('//*[@Id="blah"]')
[<Element {http://domain.com/namespace}tag at 0x7f3aee49a518>]
```

Do you have an example?

**Comment by danimaribeiro** [comment]
I see the problem now.
Adding // to the xpath finds the Id, it does not need all that code.

**Comment by kislyuk** [comment]
@soby, could you please explain the scenario where you need this?

Also, this requires a test.

**Comment by soby** [comment]
I'm using it to sign the assertion in a SAML response. The default behavior was always to sign the root of the document, which isn't always what you're looking for. I'll get the test added.

**Comment by danimaribeiro** [comment]
Could you also make this change?
https://github.com/Trust-Code/signxml/commit/7ebd988f6d23a0304e61be2a564f161136764a29

This solves my problem.

@kislyuk  With this change I can sign a xml like this, where the tag to be signed is not the root.
xml signed

**Comment by soby** [comment]
@danimaribeiro Yup, I included that. Good catch

**Comment by kislyuk** [comment]
I believe the proper way to do this should be to use `XMLSigner().sign(node, method=methods.detached)`, where `node` is the child node within the hierarchy that you want signed. This should return the detached signature node, which you can then insert where you want it. Does that not work?

**Comment by soby** [comment]
I've only ever seen enveloped signatures used in SAML. 

As a general practice, using detached signatures as enveloped signatures would probably work everywhere if you manually added the enveloped transform. That said, we still clearly have enveloped signature routines to do this for us so why not allow them to address the question of *what* to sign. It seems unnecessarily restrictive for enveloped signatures to require the entire document be included in the signature when there are very popular dsig usages such as SAML that commonly use enveloped signatures on subsets of the document.

**Comment by kislyuk** [comment]
OK, I'm convinced this is a good option to have, but I'd rather actually have it as the default if reference_uri is specified and method is not detached.

**Comment by soby** [comment]
Sounds good. I'll modify it for that.

**Comment by soby** [comment]
Hmmm...travis failed creating the test environment. Can it be re-run?

**Comment by kislyuk** [comment]
restarted

File: signxml/__init__.py [hunk 1]
```diff
@@ -365,6 +367,22 @@ def _unpack(self, data, reference_uri):
...
+                    targets = doc_root.xpath('//*[@ID="{0}"]'.format(reference_uri))
+                if len(targets) == 1:
+                    c14n_input = self.get_root(targets[0])
+                elif len(targets) == 0:
+                    raise InvalidInput("No reference found for reference_uri in enveloped signature")
+                else:
+                    raise InvalidInput("Multiple references found for reference_uri in enveloped signature")
+            else:
+                reference_uri = ""
+
```
Comment (kislyuk) [inline signxml/__init__.py]: Would you mind refactoring this to use `XMLVerifier._resolve_reference()`? (that function will also require minor refactoring to accommodate this use case)

Comment (soby) [inline signxml/__init__.py]: Done

**Comment by kislyuk** [comment]
Great, thanks for the effort. I'll roll a release shortly.

**Comment by kislyuk** [comment]
@soby, if you have any interest/time, it would be nice to include an example of how you use this to sign SAML assertions in README.rst.

**Comment by soby** [comment]
That'd look about like this but I can't for the life of me get the markdown to cooperate:

Signing SAML assertions
~~~~~~~~~~~~~~~~~~~~~~~~

 The SAML assertion schema specifies a location for the enveloped XML signature (after ``<Issuer>`` in either the Response or Assertion). To sign a SAML assertion in a schema-compliant way, insert a signature placeholder tag at that location
 before calling XMLSigner: ``<ds:Signature Id="placeholder"></ds:Signature>``.

```python
  sig_in_assertion = '''<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                                        xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="responseId">
                          <saml:Assertion xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                          xmlns:xs="http://www.w3.org/2001/XMLSchema" ID="assertionId">
                            <saml:Issuer>http://idp.example.com/metadata.php</saml:Issuer>
                            <ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#" Id="placeholder" />
                          </saml:Assertion>
                        </samlp:Response>'''
  XMLSigner().sign(etree.fromstring(sig_in_assertion), reference_uri='assertionId', key=key, cert=cert)

  sig_in_response = '''<samlp:Response xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
                                      xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion" ID="responseId">
                         <saml:Issuer>http://idp.example.com/metadata.php</saml:Issuer>
                         <ds:Signature xmlns:ds="http://www.w3.org/2000/09/xmldsig#" Id="placeholder" />
                         <saml:Assertion xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
                                        xmlns:xs="http://www.w3.org/2001/XMLSchema" ID="assertionId">
                        </saml:Assertion>
                      </samlp:Response>'''
  XMLSigner().sign(etree.fromstring(sig_in_response), reference_uri='responseId', key=key, cert=cert)
```

**Comment by kislyuk** [comment]
Released in v2.1.4, please test.

**Pull request merged**